MessageEncryptor

sbhatore95 opened the pull request.

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
     def initialize(secret, *signature_key_or_options)
-      options = signature_key_or_options.extract_options!
-      sign_secret = signature_key_or_options.first
+      @options = signature_key_or_options.extract_options!
```
Wouldn't adding `**options` to the initialize method let us get rid of `extract_options!`?

sbhatore95 replied:
This is as it was implemented before. They probably did it because we don't want passing signature key to be mandatory. I am not sure if we could pass signature key using **options. What do you say ?

kaspth replied:
We could handle that with `def initialize(secret, sign_secret = nil, **options)`

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
-      @serializer = options[:serializer] || Marshal
+      @sign_secret = signature_key_or_options.first || @secret
+      @alg = @options[:alg] || 'dir'
+      @enc = @options[:enc] || 'A128GCM'
```
Now we're using a lower bit encryption than before. What's the justification? Because it sound like something I'd rather not do :smile:

sbhatore95 replied:
Yeah, you are right. It would be better for this to be 256 bit just like previously.

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
-    def encrypt_and_sign(value)
-      verifier.generate(_encrypt(value))
+    def encrypt_and_sign(value, options = {})
+      @claims = Claims.new(payload: value, **options)
```
We don't need this to be an instance variable.

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
+          claims = _decrypt(message).symbolize_keys
+          Claims.verify!(claims, options)
+        else
+          raise InvalidFormat, 'Invalid JWT Format.'
```
Maybe `MalformedToken`, would be a better exception?

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
+      @options = signature_key_or_options.extract_options!
       @secret = secret
-      @sign_secret = sign_secret
-      @cipher = options[:cipher] || 'aes-256-cbc'
```
This breaks backwards compatibility.

sbhatore95 replied:
I don't understand how :sweat_smile: . Can you please explain ? In my understanding, backwards compatibility means that all the previously made Tokens should be handled by our new implementation. This is taken care by LegacyEncryptor https://github.com/sbhatore/rails/pull/4/files#diff-3aeeced39aff3bbba1a5c38d367eab49R179

kaspth replied:
Imagine a user upgrading to Rails 5, where they call `MessageEncryptor.new(@secret, cipher: 'aes-128-gcm')`. This won't work because there's no cipher option anymore.

sbhatore95 replied:
Oh, Now I get it!  We need the code written in old applications to work too. :+1:

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
+      end
+
+      def decrypt_content_encryption_key
+        if @alg == 'dir'
```
Why can people pass in `alg: 'something_else'` to `new`, but then not set it to anything? Just scrap this code and the other parts like it.

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
+        cipher.encrypt
+        cipher.key = content_encryption_key
+        iv = cipher.random_iv
+        cipher.iv = iv
```
Could do `cipher.iv = iv = cipher.random_iv`.

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
+        when :A256GCM
+          'aes-256-gcm'
+        else
+          raise UnexpectedAlgorithm, 'Unknown Encryption Algorithm.'
```
Why isn't the `cipher` validated when it's passed in?

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
+        cipher.auth_data = to_ascii encode(protected_header)
+
+        ciphertext = cipher.update(@serializer.dump value) + cipher.final
+        auth_tag = cipher.auth_tag
```
Inline this variable

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
+        ciphertext = cipher.update(@serializer.dump value) + cipher.final
+        auth_tag = cipher.auth_tag
+
+        ([protected_header, jwe_encrypted_key, iv, ciphertext, auth_tag].map { |a| encode a }).join('.')
```
Don't need the parens around the array + map. map returns the array, you're joining

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
-      cipher.key = @secret
-      cipher.iv  = iv
+      def cipher
+        @cipher ||= OpenSSL::Cipher.new(cipher_name)
```
I don't know if it's safe to cache the cipher. Perhaps we should look more to what I did in Tokie with the custom cipher class: https://github.com/kaspth/tokie/blob/master/lib/tokie/encryptor/cipher.rb

That keeps all the nitty gritty encryption stuff in its own place.

kaspth commented on activesupport/test/message_encryptor_test.rb:
```diff
@@ -16,31 +35,150 @@ def load(value)
+
+  def test_messing_with_message_causes_failure
+    message = @encryptor.encrypt_and_sign(@payload, @options)
+    [0, 2, 3, 4].each do |i|
```
In tests it's better to be explicit. So write these out :smile:

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -29,6 +33,9 @@ def self.dump(value)
 
     class InvalidMessage < StandardError; end
+    class UnexpectedAlgorithm < StandardError; end
+    class InvalidFormat < StandardError; end
```
Are we raising all the old exceptions in the same places as we were before? Otherwise we're breaking backwards compatibility.

sbhatore replied:
This backwards compatibility is really confusing. Does this mean that new tokens should behave in exactly the same way as the old tokens? Raise the same exceptions that were raised previously at the same places ?  I think I really need to clear my head on this :sweat:

sbhatore commented:
More tests will be added soon.

sbhatore commented:
The tests are passing but I am getting this message on running them:
```
siddharth@siddharth-Inspiron-5537:~/Development/ROR/rails/activesupport$ ruby -W -Itest test/message_encryptor_test.rb
Run options: --seed 8835

# Running:

/home/siddharth/Development/ROR/rails/activesupport/lib/active_support/encryptor/jwe.rb:7: warning: loading in progress, circular require considered harmful - /home/siddharth/Development/ROR/rails/activesupport/lib/active_support/message_encryptor.rb
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:55:in  `block in autorun'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:126:in  `run'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:149:in  `__run'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:149:in  `map'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:149:in  `block in __run'
	from /home/siddharth/Development/ROR/rails/activesupport/lib/active_support/test_case.rb:49:in  `run'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:285:in  `run'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:306:in  `with_info_handler'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:317:in  `on_signal'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:286:in  `block in run'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:286:in  `each'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:287:in  `block (2 levels) in run'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:293:in  `run_one_method'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:759:in  `run_one_method'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest/test.rb:101:in  `run'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest/test.rb:276:in  `with_info_handler'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest.rb:317:in  `on_signal'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest/test.rb:102:in  `block in run'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest/test.rb:256:in  `time_it'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest/test.rb:103:in  `block (2 levels) in run'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest/test.rb:204:in  `capture_exceptions'
	from /home/siddharth/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.3.3/lib/minitest/test.rb:104:in  `block (3 levels) in run'
	from test/message_encryptor_test.rb:30:in  `setup'
	from test/message_encryptor_test.rb:120:in  `new_encryptor'
	from /home/siddharth/Development/ROR/rails/activesupport/lib/active_support/message_encryptor.rb:8:in  `<top (required)>'
	from /home/siddharth/Development/ROR/rails/activesupport/lib/active_support/message_encryptor.rb:8:in  `require'
	from /home/siddharth/Development/ROR/rails/activesupport/lib/active_support/encryptor/agcm.rb:1:in  `<top (required)>'
	from /home/siddharth/Development/ROR/rails/activesupport/lib/active_support/encryptor/agcm.rb:1:in  `require'
	from /home/siddharth/Development/ROR/rails/activesupport/lib/active_support/encryptor/jwe.rb:6:in  `<top (required)>'
	from /home/siddharth/Development/ROR/rails/activesupport/lib/active_support/encryptor/jwe.rb:7:in  `<module:ActiveSupport>'
................

Finished in 0.016264s, 983.7409 runs/s, 2520.8359 assertions/s.

16 runs, 41 assertions, 0 failures, 0 errors, 0 skips
```
I tried removing this message but I have not been able to do so. Any idea @kaspth ?

sbhatore95 commented on Gemfile.lock:
```diff
@@ -312,6 +312,3 @@ DEPENDENCIES
   w3c_validators
-
-BUNDLED WITH
-   1.10.5
```
I have wanted to ask this for so long. Do I have to commit the Gemfile.lock file or not everytime it is updated locally ? I didn't see anyone doing it often so I didn't do it till now. But, this file changes during work.

kaspth replied:
You should upgrade to Bundler 1.10.5 or greater :)

Do `gem update bundler`.

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -1,6 +1,12 @@
+require 'active_support/core_ext/hash/keys'
+require 'active_support/claims'
+require 'active_support/message_verifier'
+require 'active_support/encryptor/agcm'
```
This is where you're getting circular require warnings.

`agcm` requires `jwe` upfront, which on line 7 mentions `class MessageEncryptor`. But this `message_encryptor.rb` file hasn't finished loading yet, so Rails' autoloading sees `ActiveSupport::MessageEncryptor` and tries to load the file we're already loading.

sbhatore95 replied:
So is it okay for these warnings to show up or is there something I can do about this ? Also, about the structure, I am not sure if we want `JWE` class to be under `MessageEncryptor` or to have a separate class for it .

kaspth commented:
I had a feeling some of these new changes were copied from somewhere else. The code that's been added doesn't read quite like what you've written before. Not to mention the rate where this amount of code has been added. Again unlike what you've demonstrated throughout this project.

These concerns made me search through other Ruby JWT implementations and found this very similar from Sandal: https://github.com/gregbeech/sandal/blob/ff965259af1572d74baab6d03fffa44b6c6e0224/lib/sandal/util.rb.
Please don't pass off other's work as your own, that's plagiarism and it's not okay.

I understand you must feel pressured to get something out the door, but this is not the way to do it.

sbhatore commented:
I have removed the AGCM and ACBC classes. I think this one is much better than previous one and it also has lesser lines of code then previous version with the same functionality. I think this PR can be opened on rails now. What do you think @kaspth ?

sbhatore95 committed:
MessageEncryptor:Authenticated Encryption of messages in JWE format

 Any message can now be encrypted as well as authenticated in JSON Web Encryption (JWE) with
 JSON Web Token (JWT) Claims set as its payload. The Authenticated Encryption with Additional
 Data (AEAD) modes implemented are AES-GCM ( Galois/Counter Mode) and AES-CBC (Cipher Block
 Chaining). The JWT implementation is as explained in MessageVerifier. Only Direct Key Encryption
 is implemented as of now. The implementation can be enhanced in future. Standard JWE spec:
 https://tools.ietf.org/html/draft-ietf-jose-json-web-encryption-40. Special thanks to
 https://github.com/gregbeech/sandal and https://github.com/nov/json-jwt for reference.

kaspth commented on activesupport/lib/active_support/encryptor/jwe.rb:
```diff
@@ -0,0 +1,157 @@
+          OpenSSL::Cipher::Cipher.new @cipher
+        end
+
+        def encryption_algorithm
```
Let's find a way to kill this entirely. It's way too much complexity for little gain.

sbhatore95 replied:
But we need to do this to adhere to the standard JWE format. `:cipher` needs to be taken care of in the options argument to maintain backwards compatibility. So, we need a way to convert `:cipher`  into the standard `enc` header parameter or wise versa and also check that the Encryption Algorithm passed in the arguments is known to us.

kaspth commented on activesupport/lib/active_support/encryptor/jwe.rb:
```diff
@@ -0,0 +1,157 @@
+          if @options[:alg].nil? || @options[:alg] == 'dir'
+            'dir'
+          else
+            raise UnexpectedAlgorithm, 'Unknown Key Encryption Algorithm.'
```
If people can't change the `alg` option they shouldn't be able to. Kill this method.

sbhatore95 replied:
:+1:

kaspth commented on activesupport/lib/active_support/encryptor/jwe.rb:
```diff
@@ -0,0 +1,157 @@
+
+        def decrypt_content_encryption_key
+          @secret
+        end
```
We definitely need to kill these methods: jwe_encrypted_key, content_encryption_key, decrypt_content_encryption_key

sbhatore95 replied:
I added these methods so that it becomes easier to add key encryption algorithms in future. For now, they seem unnecessary as I am using direct encryption. Also, the name of the functions are standard JWE names. JWE encryption key, Content Encryption Key, etc. will be very useful once I implement other key encryption algorithms.  However, it seems fair enough to keep them out until I have the implementations ready. I will kill them.

kaspth replied:
Let’s not predict the future and add only what we need now.

> Den 24/08/2015 kl. 22.34 skrev Siddharth Bhatore notifications@github.com:
> 
> In activesupport/lib/active_support/encryptor/jwe.rb https://github.com/sbhatore/rails/pull/4#discussion_r37800849:
> 
> > -          else
> > -            raise UnexpectedAlgorithm, 'Unknown Key Encryption Algorithm.'
> > -          end
> > -        end
> >   +
> > -        def jwe_encrypted_key
> > -          ''
> > -        end
> >   +
> > -        def content_encryption_key
> > -          @secret
> > -        end
> >   +
> > -        def decrypt_content_encryption_key
> > -          @secret
> > -        end
> >   I added these methods so that it becomes easier to add key encryption algorithms in future. For now, they seem unnecessary as I am using direct encryption. Also, the name of the functions are standard JWE names. JWE encryption key, Content Encryption Key, etc. will be very useful once I implement other key encryption algorithms. However, they can be removed for now if it seems necessary.
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/sbhatore/rails/pull/4/files#r37800849.

## 

Kasper

sbhatore95 replied:
:+1:

kaspth commented on activesupport/lib/active_support/encryptor/jwe.rb:
```diff
@@ -0,0 +1,157 @@
+
+        def header
+          @header ||= { 'typ' => 'JWT', 'alg' => @alg.to_s, 'enc' => @enc.to_s }.tap do |header|
+            header['kid'] = @options[:kid] if @options[:kid]
```
What does kid mean again?

sbhatore95 replied:
https://tools.ietf.org/html/draft-ietf-jose-json-web-encryption-40#section-4.1.6
It has been suggested to add by Tony and also discussed in the previous basecamp threads.

kaspth commented on activesupport/lib/active_support/encryptor/jwe.rb:
```diff
@@ -0,0 +1,157 @@
+        end
+
+        def digest
+          "sha#{@enc.split(//).last(3).join}"
```
Why do we need this?

sbhatore95 replied:
It is used in `generate_digest` method.

kaspth replied:
I don’t think ”it’s used in another method” is a sufficient justification :grin: What’s its purpose? It looks like a lot of mangling to me.

> Den 24/08/2015 kl. 22.29 skrev Siddharth Bhatore notifications@github.com:
> 
> In activesupport/lib/active_support/encryptor/jwe.rb https://github.com/sbhatore/rails/pull/4#discussion_r37800325:
> 
> > -        def content_encryption_key
> > -          @secret
> > -        end
> >   +
> > -        def decrypt_content_encryption_key
> > -          @secret
> > -        end
> >   +
> > -        def header
> > -          @header ||= { 'typ' => 'JWT', 'alg' => @alg.to_s, 'enc' => @enc.to_s }.tap do |header|
> > -            header['kid'] = @options[:kid] if @options[:kid]
> > -          end
> > -        end
> >   +
> > -        def digest
> > -          "sha#{@enc.split(//).last(3).join}"
> >   It is used in generate_digest method.
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/sbhatore/rails/pull/4/files#r37800325.

## 

Kasper

sbhatore95 replied:
Okay. So we have two algorithms for AEAD encryption which need this method. A256CBC-HS512 and A128CBC-HS256. Notice that signing algorithms used are SHA-512 and SHA-256 respectively. So this method takes the last three characters to give the algorithm used in `generate_digest` method :smile: .

kaspth commented on activesupport/lib/active_support/encryptor/jwe.rb:
```diff
@@ -0,0 +1,157 @@
+        @serializer = @options[:serializer] || Marshal
+      end
+
+      def encrypt(value)
```
Both encrypt and decrypt are still extremely tough to understand. We need to find a way to make them simpler.

kaspth commented on activesupport/lib/active_support/encryptor/jwe.rb:
```diff
@@ -0,0 +1,157 @@
+        @secret = @options[:secret]
+        @alg = key_encryption_algorithm
+        @cipher = encryption_algorithm
+        @enc = CIPHER_ENC_PAIRS[@cipher]
```
Let's kill `@enc` completely, the `@cipher` tells us what we need to know.

sbhatore95 replied:
:+1:

kaspth commented:
It still feels like there's a lot of code that doesn't need to be there. It's adding way too much complexity for little gain.

I hope I haven't been too harsh - I'm a little sleepy now. Keep working on it, I'm sure you'll get there.

kaspth commented:
@sbhatore My comments haven't been addressed, are you still up for working on this?

If you're pressed for time (and I'd like to get this into Rails 5) perhaps we can do a split, where you handle the simpler message verifier and I'll take over the message encryptor part? :grin:
